transport-native-unix-common/src/main/java/io/netty/channel/unix/DirectIoByteBufPool.java

```diff
@@ -0,0 +1,1312 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.unix;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.DuplicatedByteBuf;
+import io.netty.buffer.ReadOnlyByteBuf;
+import io.netty.buffer.SlicedByteBuf;
+import io.netty.buffer.SwappedByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.util.ByteProcessor;
+import io.netty.util.IllegalReferenceCountException;
+import io.netty.util.concurrent.FastThreadLocal;
+import io.netty.util.internal.SystemPropertyUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.ScatteringByteChannel;
+import java.nio.charset.Charset;
+
+import static io.netty.util.internal.ObjectUtil.checkPositive;
+
+final class DirectIoByteBufPool {
+
+    private static final int MAX_CACHED_SIZE = SystemPropertyUtil.getInt(
+            "jdk.nio.maxCachedBufferSize", Integer.MAX_VALUE); // Use the same property name as the JDK.
+
+    // We need the Cleaner to ensure the memory will eventually be released as the user may not use a
+    // FastThreadLocalThread.
+    private static final ByteBufAllocator IO_BUFFER_ALLOCATOR = new UnpooledByteBufAllocator(true, false, false);
+
+    private static final FastThreadLocal<DirectIoByteBufPool> POOL = new FastThreadLocal<DirectIoByteBufPool>() {
+        @Override
+        protected DirectIoByteBufPool initialValue() {
+            return new DirectIoByteBufPool(Thread.currentThread());
+        }
+
+        @Override
+        protected void onRemoval(DirectIoByteBufPool value) {
+            value.free();
+        }
+    };
+
+    private ByteBuf parent;
+
+    // Just used to assert for the correct thread when releasing the buffer back to the pool.
+    private final Thread ioThread;
+
+    private DirectIoByteBufPool(Thread ioThread) {
+        this.ioThread = ioThread;
+        parent = IO_BUFFER_ALLOCATOR.directBuffer(4096);
+    }
+
+    static ByteBuf acquire(int capacity) {
+        ByteBuf buffer = POOL.get().get(capacity);
+        return buffer != null ? buffer : IO_BUFFER_ALLOCATOR.directBuffer(capacity);
+    }
+
+    private ByteBuf get(int reqCapacity) {
+        if (parent.capacity() + reqCapacity > MAX_CACHED_SIZE) {
+            return null;
+        }
+
+        // No space left...
+        if (parent.writableBytes() < reqCapacity) {
+            int capacity = IO_BUFFER_ALLOCATOR.calculateNewCapacity(parent.capacity() + reqCapacity,
+                    parent.capacity() + reqCapacity * 2);
+            parent = IO_BUFFER_ALLOCATOR.directBuffer(capacity);
+        }
+
+        int writerIndex = parent.writerIndex();
+        // TODO: Improve GC if possible
```

Comment:
How about: Instead of defining a brand new `PooledIoByteBuf` type used for every sub-buffer, just wrap `parent` in some `AbstractDerivedByteBuf` subclass and override the `refCnt0` etc methods to use a non-volatile field (I know they are package-private,could maybe change them to `protected` or work around that somehow).

That will have additional advantage of avoiding volatile update when making the `retainedSlice`s below (and corresponding later release).